extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/devconsole/DevConsoleProcessor.java

```diff
@@ -32,61 +31,78 @@ public DevConsoleRuntimeTemplateInfoBuildItem collectScores() {
+        return new DevConsoleTemplateInfoBuildItem("staticResourceInfo", staticResourceInfo);
     }
 
-    private void collectKnownPaths(Path resource, Set<String> staticResources) {
+    private void collectKnownPaths(Path resource, StaticResourceInfo staticResourceInfo) {
+
         try {
             Files.walkFileTree(resource, new SimpleFileVisitor<Path>() {
+                final Stack<StaticResourceInfo.StaticFile> currentFolder = new Stack<>();
 
                 @Override
-                public FileVisitResult visitFile(Path p, BasicFileAttributes attributes) throws IOException {
-                    String simpleName = p.getFileName().toString();
-                    String file = resource.relativize(p).toString();
-                    if (simpleName.equals("index.html") || simpleName.equals("index.htm")) {
-                        Path parent = resource.relativize(p).getParent();
-                        if (parent == null) {
-                            staticResources.add("/");
-                        } else {
-                            String parentString = parent.toString();
-                            if (!parentString.startsWith("/")) {
-                                parentString = "/" + parentString;
-                            }
-                            staticResources.add(parentString + "/");
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+
+                    Path relativeFile = resource.relativize(dir);
+                    String relativeFileString = "/" + relativeFile.toString();
+                    Path dirName = dir.getFileName();
+                    StaticResourceInfo.StaticFile getStaticFile = staticResourceInfo.resourceMap.get(relativeFileString);
+
+                    if (getStaticFile == null) {
+                        getStaticFile = new StaticResourceInfo.StaticFile(dirName.toString(), true);
+                        staticResourceInfo.resourceMap.put(relativeFileString, getStaticFile);
+
+                        // adding to parent
+                        if (!currentFolder.isEmpty()) {
+                            currentFolder.peek().children.add(getStaticFile);
                         }
+                        currentFolder.push(getStaticFile);
                     }
-                    if (!file.startsWith("/")) {
-                        file = "/" + file;
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    String fileName = file.getFileName().toString();
+
+                    StaticResourceInfo.StaticFile newStaticFile = new StaticResourceInfo.StaticFile(fileName, false);
+
+                    if (!currentFolder.empty() && !currentFolder.peek().children.contains(newStaticFile)) {
```

Comment:
CurrentFolder can't be empty, I don't think. And here you're trying `contains` but it relies on `equals` to be implemented, which it's not, so it will revert to reference equality, so it will never return `true` because you just created the file.

I don't think there can be file duplication, so just don't care about this for now :)

If you must, then traverse `children` and look at the file names.